File: file-service/src/main/java/net/blueshell/fileservice/controller/FileController.java
package net.blueshell.fileservice.controller;

import net.blueshell.common.dto.FileDTO;
import net.blueshell.db.BaseController;
import net.blueshell.fileservice.mapper.FileMapper;
import net.blueshell.fileservice.service.FileService;
import net.blueshell.fileservice.model.File;
import net.blueshell.fileservice.repository.FileRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Controller
public class FileController extends BaseController<FileService, FileRepository> {

    private final FileMapper fileMapper;

    @Autowired
    public FileController(FileService service, FileRepository repository, FileMapper fileMapper) {
        super(service, repository);
        this.fileMapper = fileMapper;
    }

    @GetMapping("/{filename:.+}")
    @ResponseBody
    public ResponseEntity<Resource> downloadFile(@PathVariable String filename) {
        File file = service.findByName(filename);
        return service.prepareFileResponse(file);
    }

    @PostMapping
    public List<FileDTO> uploadFile(List<FileDTO> dtos) {
        List<File> files = fileMapper.fromDTOs(dtos);
        return fileMapper.toDTOs(files);
    }
}

File: services/api/src/main/kotlin/net/blueshell/api/domain/file/web/FileController.kt
package net.blueshell.api.domain.file.web

import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.constraints.NotNull
import net.blueshell.api.domain.file.application.FileService
import net.blueshell.api.shared.enums.FileType
import net.blueshell.api.domain.file.web.dto.response.FileResponse
import net.blueshell.api.domain.file.web.mapping.response.asResponse
import net.blueshell.api.domain.file.web.validation.AllowedContentTypes
import net.blueshell.api.domain.file.web.validation.FileSize
import net.blueshell.api.shared.web.BaseController
import org.springframework.core.io.Resource
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile

@RestController
@Tag(name = "Files")
class FileController(
    service: FileService
) : BaseController<FileService>(service) {
    @GetMapping("/events/{eventId}/banners")
    @PreAuthorize("hasPermission(#eventId, 'Event', 'read')")
    fun downloadEventBanner(@PathVariable eventId: Long): ResponseEntity<Resource> {
        return service.prepareFileResponse(service.findByBannerEventId(eventId))
    }

    @PostMapping(value = ["/events/banners"], consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
    @ResponseStatus(
        HttpStatus.CREATED
    )
    @PreAuthorize("hasPermission('__NO_TARGET__', 'EventBanner', 'write')")
    fun uploadEventBanner(
        @RequestPart("file") @NotNull(message = "File is required") @FileSize(max = 10 * 1024 * 1024) @AllowedContentTypes(
            "image/png",
            "image/jpeg",
            "image/jpg",
            "image/webp",
            "image/gif"
        ) file: @NotNull(message = "File is required") MultipartFile
    ): FileResponse {
        return service.storeMultipart(file, FileType.EVENT_BANNER).asResponse()
    }
}
